Iterable-iOS-SDK/IterableAPI.m

```diff
@@ -223,18 +223,39 @@ - (void)sendRequest:(NSURLRequest *)request onSuccess:(void (^)(NSDictionary *))
                                                                    NSURLResponse *response,
                                                                    NSError *error)
     {
-        if ([data length] > 0 && error == nil) {
-            error = nil;
-            id object = [NSJSONSerialization
-                         JSONObjectWithData:data
-                         options:0
-                         error:&error];
-            if(error) {
+        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
+        NSInteger responseCode = httpResponse.statusCode;
+        
+        NSError *jsonError = nil;
+        id json = [NSJSONSerialization
+                     JSONObjectWithData:data
+                     options:0
+                     error:&jsonError];
+
+        if (responseCode == 401) {
+            if (onFailure != nil) onFailure(@"Invalid API Key", data);
+        }
+        else if (responseCode >= 400) {
+            NSString *errorMessage = @"Invalid Request";
+            if (responseCode >= 500) {
+                errorMessage = @"Internal Server Error";
+            }
+
+            // Extract error message from the response JSON
+            if ([json isKindOfClass:[NSDictionary class]] && [json objectForKey:@"msg"] &&
+                    [[json objectForKey:@"msg"] isKindOfClass:[NSString class]]) {
+                errorMessage = [json objectForKey:@"msg"];
+            }
+
+            if (onFailure != nil) onFailure(errorMessage, data);
+        }
+        else if (responseCode == 200 && [data length] > 0 && error == nil) {
```

Comment:
do we necessarily need to check for `responseCode == 200`? what happens for redirects? it follows the redirect and returns the final code right?

if you do end up wanting to keep the explicit check on `200`, you should probably split the other case up, like
```
if ([data length] > 0 && error == nil) {
  if (responseCode == 200) {
    // stuff you have now
  } else {
    if (onFailure != nil) onFailure("received non-200 response @responseCode", data);
  }
}
```

Reply:
Does API ever return redirects?

Reply:
it seems the only api that returns redirects atm is the view in browser api, which isn't implemented in the sdk

the other thing which definitely returns redirects is in-app links... not quite where/how those are handled

Reply:
In-app links that are inside in-apps' HTML? IIRC those links are not API URLs.

Reply:
yeah looks like it's done in a separate code path here:
https://github.com/Iterable/iterable-ios-sdk/blob/6577b4c49c4338147d6b9ae5be39b20b41404efd/Iterable-iOS-SDK/IterableAPI.m#L417
https://github.com/Iterable/iterable-ios-sdk/blob/c4aff0074b23ccb6eb93631a16e62babb30606de/Iterable-iOS-SDK/IterableDeeplinkManager.m#L45

so i guess it's ok to check for 200